src/modinfo.c

```diff
@@ -3892,36 +3934,39 @@ sr_modinfo_push_oper_mod_learn_changes(sr_session_ctx_t *sess, const char *mod_n
 }
 
 /**
- * @brief Update push oper mod data in a session, add a new module if not yet present.
+ * @brief Update push oper mod data cache in the session, add new module if not yet present.
  *
  * @param[in] sess Session to update.
  * @param[in] mod_name Module name.
- * @param[in] has_data Flag to store.
+ * @param[in] data module data to store.
  * @return err_info, NULL on success, only fails if out of memory.
  */
 static sr_error_info_t *
-sr_modinfo_push_oper_mod_update_sess(sr_session_ctx_t *sess, const char *mod_name, int has_data)
+sr_modinfo_push_oper_mod_update_cache(sr_session_ctx_t *sess, const char *mod_name, struct lyd_node *data)
 {
     sr_error_info_t *err_info = NULL;
     uint32_t i;
     void *mem;
 
     for (i = 0; i < sess->oper_push_mod_count; ++i) {
         if (!strcmp(sess->oper_push_mods[i].name, mod_name)) {
-            sess->oper_push_mods[i].has_data = has_data;
-            return NULL;
+            break;
         }
     }
 
-    /* add new module */
-    mem = realloc(sess->oper_push_mods, (i + 1) * sizeof *(sess->oper_push_mods));
-    SR_CHECK_MEM_GOTO(!mem, err_info, cleanup);
-    sess->oper_push_mods = mem;
+    if (i == sess->oper_push_mod_count) {
+        /* add new module */
+        mem = realloc(sess->oper_push_mods, (i + 1) * sizeof *(sess->oper_push_mods));
+        SR_CHECK_MEM_GOTO(!mem, err_info, cleanup);
+        sess->oper_push_mods = mem;
+
+        sess->oper_push_mods[i].name = strdup(mod_name);
+        SR_CHECK_MEM_GOTO(!sess->oper_push_mods[i].name, err_info, cleanup);
+        ++sess->oper_push_mod_count;
+    }
 
-    sess->oper_push_mods[i].name = strdup(mod_name);
-    SR_CHECK_MEM_GOTO(!sess->oper_push_mods[i].name, err_info, cleanup);
-    sess->oper_push_mods[i].has_data = has_data;
-    ++sess->oper_push_mod_count;
+    sess->oper_push_mods[i].has_data = !!data;
+    sess->oper_push_mods[i].cache = data;
```

Comment:
Is it guaranteed that `cache` is `NULL`? Well, obviously not when adding a new module because `realloc` is used but must the data be spent otherwise? If so, I would add a comment and assert at 3953.

Reply:
Yes, it should be guaranteed to be `NULL`, after the first `sr_modinfo_get_oper_data()` call in `sr_apply_oper_changes()`.

```C
    /* get the operational DS data with the old push oper data */
    if ((err_info = sr_modinfo_get_oper_data(mod_info, session, NULL))) {
        goto cleanup;
    }
```

Since operational datastore does not support factory reset RPC, or `sr_replace_config()`, `sr_modinfo_data_store()` should never be called for a module with its cache still present.